scripts/generate-docs.go

```diff
@@ -0,0 +1,307 @@
+// 9. replace /templates/{data-sources/resources} directories with their backed up equivalents
+//
+package main
+
+import (
+	"fmt"
+	"html/template"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+// PageData represents a type of service and enables a template to render
+// different content depending on the service type.
+type PageData struct {
+	ServiceType string
+}
+
+// Page represents a template page to be rendered.
+//
+// name: the {{ define "..." }} in each template.
+// path: the path to write rendered output to.
+// Data: context specific information (e.g. vcl vs wasm).
+//
+// Data is public as it's called via the template processing logic.
+type Page struct {
+	name string
+	path string
+	Data PageData
+}
+
+func main() {
+	baseDir := getBaseDir()
+	tmplDir := baseDir + "/templates"
+
+	tempDir, err := ioutil.TempDir("", "precompile")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.RemoveAll(tempDir)
+
+	var dataPages = []Page{
+		{
+			name: "ip_ranges",
+			path: tempDir + "/data-sources/ip_ranges.md.tmpl",
+		},
+		{
+			name: "waf_rules",
+			path: tempDir + "/data-sources/waf_rules.md.tmpl",
+		},
+	}
+
+	var resourcePages = []Page{
+		{
+			name: "service_v1",
+			path: tempDir + "/resources/service_v1.md.tmpl",
+			Data: PageData{
+				"vcl",
+			},
+		},
+		{
+			name: "service_compute",
+			path: tempDir + "/resources/service_compute.md.tmpl",
+			Data: PageData{
+				"wasm",
+			},
+		},
+		{
+			name: "service_dictionary_items_v1",
+			path: tempDir + "/resources/service_dictionary_items_v1.md.tmpl",
+		},
+		{
+			name: "service_acl_entries_v1",
+			path: tempDir + "/resources/service_acl_entries_v1.md.tmpl",
+		},
+		{
+			name: "service_dynamic_snippet_content_v1",
+			path: tempDir + "/resources/service_dynamic_snippet_content_v1.md.tmpl",
+		},
+		{
+			name: "service_waf_configuration",
+			path: tempDir + "/resources/service_waf_configuration.md.tmpl",
+		},
+		{
+			name: "user_v1",
+			path: tempDir + "/resources/user_v1.md.tmpl",
+		},
+	}
+
+	pages := append(resourcePages, dataPages...)
+
+	renderPages(getTemplate(tmplDir), pages)
+
+	appendSyntaxToFiles(tempDir)
+
+	copyIndexToTempDir(tmplDir, tempDir)
+
+	backupTemplatesDir(tmplDir)
+
+	replaceTemplatesDir(tmplDir, tempDir)
+
+	runTFPluginDocs()
+
+	replaceTemplatesDir(tmplDir, tmplDir+"-backup")
+}
+
+// getBaseDir returns the terraform repo directory.
+//
+// TODO: simplify this logic with something like:
+// git rev-parse --show-toplevel
+func getBaseDir() string {
+	_, scriptPath, _, ok := runtime.Caller(0)
+	if !ok {
+		log.Fatal("Could not get current working directory")
+	}
+	for !strings.HasPrefix(filepath.Base(scriptPath), "terraform-provider-") && scriptPath != "/" {
+		scriptPath = filepath.Clean(scriptPath + "/..")
+	}
+	if scriptPath == "/" {
+		log.Fatal("Script was run outside of fastly provider directory")
+	}
+	return scriptPath
+}
```

Comment:
I'm unsure why this is needed? Wouldn't [`os.Getwd()`](https://golang.org/pkg/os/#Getwd) be sufficient, if we know this script is invoked for a make target in the project root directory?